relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,225 @@
+#[derive(Debug)]
+pub struct GlobalConfigService {
+    /// Sender of the [`watch`] channel for the subscribers of the service.
+    sender: watch::Sender<Arc<GlobalConfig>>,
+    /// Sender of the internal channel to forward global configs from upstream.
+    internal_tx: UnboundedSender<Result<GetGlobalConfigResponse, UpstreamRequestError>>,
+    /// Receiver of the internal channel to forward global configs from upstream.
+    internal_rx: UnboundedReceiver<Result<GetGlobalConfigResponse, UpstreamRequestError>>,
+    /// Upstream service to request global configs from.
+    upstream: Addr<UpstreamRelay>,
+    /// The duration to wait before fetching global configs from upstream.
+    fetch_interval: Duration,
+    /// Handle to avoid multiple outgoing requests.
+    fetch_handle: SleepHandle,
+}
+
+impl GlobalConfigService {
+    /// Creates a new [`GlobalConfigService`].
+    pub fn new(config: Arc<Config>, upstream: Addr<UpstreamRelay>) -> Self {
+        let (sender, _) = watch::channel(Arc::default());
+        let (internal_tx, internal_rx) = mpsc::unbounded_channel();
+        Self {
+            sender,
+            internal_tx,
+            internal_rx,
+            upstream,
+            fetch_interval: config.global_config_fetch_interval(),
+            fetch_handle: SleepHandle::idle(),
+        }
+    }
+
+    /// Handles messages from external services.
+    fn handle_message(&self, message: GlobalConfigManager) {
+        match message {
+            GlobalConfigManager::Get(sender) => {
+                sender.send(self.sender.borrow().clone());
+            }
+            GlobalConfigManager::Subscribe(sender) => {
+                sender.send(self.sender.subscribe());
+            }
+        }
+    }
+
+    /// Schedules the next global config request.
+    fn schedule_fetch(&mut self) {
+        if self.fetch_handle.is_idle() {
+            // XXX(iker): set a backoff mechanism?
+            self.fetch_handle.set(self.fetch_interval);
+        }
+    }
+
+    /// Requests a new global config from upstream.
+    ///
+    /// As the upstream service is responsible for authentication, the global
+    /// config service doesn't deal with it.
+    fn update_global_config(&mut self) {
+        self.fetch_handle.reset();
+
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let internal_tx = self.internal_tx.clone();
+
+        tokio::spawn(async move {
+            let query = GetGlobalConfig::query();
+            match upstream_relay.send(SendQuery(query)).await {
+                // TODO(iker): add observability on the request.
+                Ok(res) => {
+                    if internal_tx.send(res).is_err() {
+                        relay_log::error!("failed to forward internally the upstream response");
+                    }
+                }
+                // FIXME(iker): if the request to upstream fails, we should
+                // reschedule another fetch anyway.
```

Comment:
Other Relay services have `SleepHandle`s and `RetryBackoff`s. These services use the backoff mechanism to know if a request was started, but they also assume a frequent number of incoming messages and use that chance to check the backoff. The global config service can't rely on that assumption due to the lower number of incoming messages, so I see two options here.

1. Accept more than one ongoing request at a time (not amazing).
2. Accept that no more global config updates will be sent if an upstream request fails (pretty bad).

Would love to see if somebody else has a different opinion or suggestions.

Reply:
I think I solved it by passing a double result to `handle_upstream_response`. now it will also handle the case for when the sending itself fails, so that new global configs will still be fetched.  
 
Now it will only fail if the internal receiver is dropped before the sender, which should be impossible since they both have the same lifetime.